general/kdtree.hpp

```diff
@@ -0,0 +1,765 @@
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_KDTREE_HPP
+#define MFEM_KDTREE_HPP
+
+#include <vector>
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+#include <cmath>
+
+#include "../linalg/vector.hpp"
+
+namespace mfem
+{
+
+namespace KDTreeNorms
+{
+
+/// Evaluates l1 norm of a vector.
+template <typename Tfloat, int ndim>
+struct Norm_l1
+{
+   Tfloat tm;
+   Tfloat operator() (const Tfloat* xx)
+   {
+      if (xx[0]<Tfloat(0.0)) { tm=-xx[0];}
+      else { tm=xx[0];}
+      for (int i=1; i<ndim; i++)
+      {
+         if (xx[i]<Tfloat(0.0)) { tm=tm-xx[i]; }
+         else { tm=tm+xx[i];}
+      }
+      return tm;
+   }
+};
+
+/// Evaluates l2 norm of a vector.
+template<typename Tfloat,int ndim>
+struct Norm_l2
+{
+   Tfloat tm;
+   Tfloat operator() (const Tfloat* xx)
+   {
+      tm=xx[0]*xx[0];
+      for (int i=1; i<ndim; i++)
+      {
+         tm=tm+xx[i]*xx[i];
+      }
+      return sqrt(tm);
+   }
+};
+
+/// Finds the max absolute value of a vector.
+template<typename Tfloat,int ndim>
+struct Norm_li
+{
+   Tfloat tm;
+   Tfloat operator() (const Tfloat* xx)
+   {
+      if (xx[0]<Tfloat(0.0)) { tm=-xx[0];}
+      else { tm=xx[0];}
+      for (int i=1; i<ndim; i++)
+      {
+         if (xx[i]<Tfloat(0.0))
+         {
+            if (tm<(-xx[i])) {tm=-xx[i];}
+         }
+         else
+         {
+            if (tm<xx[i]) {tm=xx[i];}
+         }
+      }
+      return tm;
+   }
+};
+
+}
+
+/// Template class for build KDTree with template parameters Tindex
+/// specifying the type utilized for indexing the points, Tfloat
+/// specifying a float type for representing the coordinates of the
+/// points, integer parameter ndim specifying the dimensionality of the
+/// space and template function Tnorm for evaluating the distance
+/// between two points.
+template <typename Tindex, typename Tfloat, size_t ndim=3, typename Tnorm=KDTreeNorms::Norm_l2<Tfloat,ndim> >
+class KDTree
+{
+public:
+
+   /// Structure defining a geometric point in the ndim-dimensional
+   /// space.
+   struct PointND
+   {
+
+      /// Geometric point constructor
+      PointND() {for (size_t i=0; i<ndim; i++) {xx[i]=Tfloat(0.0);}}
+
+      /// Coordinates of the point
+      Tfloat xx[ndim];
+   };
+
+   /// Structure defining a node in the KDTree.
+   struct NodeND
+   {
+
+      /// Defines a point in the ndim-dimensional space
+      PointND pt;
+
+      /// Defines the attached index
+      Tindex  ind;
+   };
+
+   /// Default constructor
+   KDTree() {}
```

Comment:
I want to add several constructors in order to avoid the dynamic memory allocation necessary for building the tree. However, I would prefer first to fix the functionality.